assets/js/src/core/hooks/use-api-error-handler.tsx

```diff
@@ -0,0 +1,58 @@
+/**
+* Pimcore
+*
+* This source file is available under two different licenses:
+* - Pimcore Open Core License (POCL)
+* - Pimcore Commercial License (PCL)
+* Full copyright and license information is available in
+* LICENSE.md which is distributed with this source code.
+*
+*  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+*  @license    https://github.com/pimcore/studio-ui-bundle/blob/1.x/LICENSE.md POCL and PCL
+*/
+
+import { useEffect } from 'react'
+import { type FetchBaseQueryError } from '@reduxjs/toolkit/query'
+import { type SerializedError } from '@reduxjs/toolkit'
+import { isEmpty, isString } from 'lodash'
+import { useAlertModal } from '@Pimcore/components/modal/alert-modal/hooks/use-alert-modal'
+
+type UseApiErrorHandler = (errorData: FetchBaseQueryError | SerializedError | undefined | null) => void
+
+interface IErrorData {
+  detail: string
+  message: string
+}
+
+const DEFAULT_ERROR_CONTENT = 'Something went wrong.'
+
+export const useApiErrorHandler: UseApiErrorHandler = (errorData) => {
```

Comment:
I think we should design this a bit more generic. 
The odds are not to small that we want to report other Errors, than just APIErrors in the future. 

Like mentioned in the code sample I would report real errors. 
And based on the type `instanceof error = APIError` we can handle the different cases.
That would even cover inheritance and would allow us to even define different behaviors for different error types.

Or we could even open this further up, to also allow the community to define there own error types and handling.

Reply:
Yes we should have different types of errors and all errors should be catched centrally in the error boundary and based on the type of error the boundary should decide how to handle/display them. But one use hook specialized for the api errors like this could in my opinion still make sense, wouldn't it? At the end this hook could also trigger/throw the regular error types but it would just be an helper to create the right (default) error type with the right (default) content.

Reply:
@vin0401
Could you elaborate in more detail, please? What other errors do you mean?

From my perspective, it still seems beneficial to have a separate hook dedicated only to API errors, with the flexibility to update it for any new requirements in the future. In my opinion, the main drawback of having common hook for all types of errors is the potential for type conflicts. Given that the API is a significant part of the project, this approach seems to introduce unnecessary complexity.

Currently, I see two options: handling errors from the API separately and catching all others in the ErrorBoundary. To me, this approach seems to be enough.

What is your opinion?

Reply:
In general I'm fine with having a separate handler for the APIErrors. 
But these are not the only errors that can happen. 

For example we could have a valid response from the API but there is a piece of information missing - maybe in some kind of configuration that prevents us from processing whatever we are trying to do at that moment. For sure there are also other cases. And for that it would be good to have a general error handler in hand that allows us to handle these errors. Or even better we would have the possibility to describe how a error should be handled by this global error handler. 

This is just a rough idea so don't take it to serios just want to point out what I have in mind. :D 

``` typescript
abstract class AbstractErrorHandler {
    processibleErrors: (typeOf Error)[]
    handleError(error) {}
    // ...
} 

class APIErrorHandler extends AbstactErrorHandler {
  processibleErrors: [APIError]
  
  handleError() {
    // describe handling of API Errors
  }
}

class ErrorProcessor {
  errorHandlers: AbstractErrorHanlder[]
  
  // register a new errorHandler
  register(errorHanlder) {}
  
  // check for a matching error handler and process the error
  process(error) {
    const errorHandler = errorHanlders.find(handler => handler.processibleErrors.includes(instanceOf error))
    
    errorHandler.handleError(error);
  }
}

// ----
// somewhere at the early stage of the Application
errorProcessor.register(new APIErrorHanlder());
// ----

const useErrorHandler = () => {   
   report(error) {
     throw new error;
  }
}

class ErrorBoundary extends React.Component {
  // catch the error via react Error Boundary or via window.on('error', {}) depending on the origin of the error either react rendering issue or async event. And process it via ErrorProcessor.process(error)
}
```

That way we would have the possibility to describe the handling for all kind of different Errors and on the other hand could also handle unknown ones or just rethrow them. We could even think of adding a priority in the future, in case we have ErrorHandlers that handle the same ErrorTypes.

Reply:
Thank you for so detail explaining, I catch your thoughts and on general like it :) However, I still have concern that there may not be many cases for the type of errors you described as examples and that we will have the same behavior how to handle it. For this reason, I'm not sure it makes sense to expand the logic as I would prefer to implement global logic only when there is a common need, such as handling API errors.

To bottom line, I would like to suggest in current implementation go with the logic of having only API error handling. However, if we see in the future that we need to handle other types of errors and this handling should be consistent, let's expand the approach.

@vin0401 @markus-moser What are your opinion?

Reply:
We already have a set of errors which are already thrown:
https://github.com/search?q=repo%3Apimcore%2Fstudio-ui-bundle+throw&type=code&p=1

These also need to be shown to the user. 
Most of them will end up at the error boundary, but still we want to control the behavior of these Errors in the the boundary and the errorHandler would be the best approach for that in my opinion. 

So having this in place would be really useful and would prevent us on searching for custom solutions that all work pretty similar. 

I'm open for other approaches but I definitely see the need to unify the logic for these.

Reply:
@vin0401 @markus-moser 
I'll take in work the last point from Marvin, and it generally looks like we are on the right track. Thanks for sharing your ideas and thoughts, a real teamwork vibe 😍

Reply:
I totally agree with the suggested ideas of @vin0401. As it is a really important topic we should invest the time to make it flexible to be able to work for various types of errors.